src/psyclone/tests/psyad/transformations/test_preprocess.py

```diff
@@ -158,10 +165,116 @@ def test_preprocess_arrayrange2loop(tmpdir, fortran_reader, fortran_writer):
+        "  a = b * c + b * d\n\n"
+        "end program test\n")
+    psyir = fortran_reader.psyir_from_source(code)
+    preprocess_trans(psyir, ["a", "c", "d"])
+    result = fortran_writer(psyir)
+    assert result == expected
+
+
+@pytest.mark.parametrize("operation", ["+", "-"])
+def test_associativity1(operation, fortran_reader, fortran_writer):
+    '''Test that the associativity function works as expected for x*(a+-b)
+    where a and b are active and x is inactive.
+
+    '''
+    code = (
+        f"program test\n"
+        f"  integer :: a, b, c, d, e, f\n"
+        f"  a = (b*e/f) * (c {operation} d)\n"
+        f"end program test\n")
+    expected = (
+        f"program test\n"
+        f"  integer :: a\n"
+        f"  integer :: b\n"
+        f"  integer :: c\n"
+        f"  integer :: d\n"
+        f"  integer :: e\n"
+        f"  integer :: f\n\n"
+        f"  a = b * e / f * c {operation} b * e / f * d\n\n"
+        f"end program test\n")
+    psyir = fortran_reader.psyir_from_source(code)
+    associativity(psyir.children[0][0], ["a", "c", "d"])
+    result = fortran_writer(psyir)
+    assert result == expected
+
+
+@pytest.mark.parametrize("operation", ["*", "/"])
+def test_associativity2(operation, fortran_reader, fortran_writer):
+    '''Test that the associativity function works as expected for (a+b)*/x
+    where a and b are active and x is inactive.
+
+    '''
+    code = (
+        f"program test\n"
+        f"  integer :: a, b, c, d, e, f\n"
+        f"  a = (c + d) {operation} (b*e/f)\n"
+        f"end program test\n")
+    expected = (
+        f"program test\n"
+        f"  integer :: a\n"
+        f"  integer :: b\n"
+        f"  integer :: c\n"
+        f"  integer :: d\n"
+        f"  integer :: e\n"
+        f"  integer :: f\n\n"
+        f"  a = c {operation} (b * e / f) + d {operation} (b * e / f)\n\n"
+        f"end program test\n")
+    psyir = fortran_reader.psyir_from_source(code)
+    associativity(psyir.children[0][0], ["a", "c", "d"])
+    result = fortran_writer(psyir)
+    assert result == expected
+
+
+def test_associativity3(fortran_reader, fortran_writer):
+    '''Test that the associativity function works as expected when the
+    function needs to be applied multiple times in an assoignment.
+
+    '''
+    code = (
+        "program test\n"
+        "  integer :: a, b, c, d, e, f, g\n"
+        "  a = b * (c + f * (d + g)) / e\n"
+        "end program test\n")
+    expected = (
+        "program test\n"
+        "  integer :: a\n"
+        "  integer :: b\n"
+        "  integer :: c\n"
+        "  integer :: d\n"
+        "  integer :: e\n"
+        "  integer :: f\n"
+        "  integer :: g\n\n"
+        "  a = b * c / e + (b * (f * d) / e + b * (f * g) / e)\n\n"
+        "end program test\n")
+    psyir = fortran_reader.psyir_from_source(code)
+    associativity(psyir.children[0][0], ["d", "c", "g"])
+    result = fortran_writer(psyir)
+    assert result == expected
```

Comment:
Please add a test for the case where the expression contains array ranges.

Reply:
Added.